lib/future.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+class Future {
+  constructor(executor) {
+    this.executor = executor;
+  }
+
+  static of(value) {
+    return new Future(resolve => resolve(value));
+  }
+
+  static err(error) {
+    return new Future((resolve, reject) => reject(error));
+  }
+
+  chain(fn) {
+    return new Future((resolve, reject) =>
+      this.run(value => fn(value).run(resolve, reject), error => reject(error))
+    );
+  }
+
+  map(fn) {
+    return this.chain(
+      value =>
+        new Future((resolve, reject) => {
+          try {
+            resolve(fn(value));
+          } catch (error) {
+            reject(error);
+          }
+        })
+    );
+  }
```

Comment:
Could this perhaps be rewritten without double Future?
```js
  map(fn) {
    return new Future((resolve, reject) => {
      this.run(value => {
        try {
          resolve(fn(value));
        } catch (error) {
          reject(error);
        }
      });
    });
```

Reply:
Thanks for bringing the attention here!
I was able to spot the inconsistency of error handling.
The ideologically correct way to implement `map` function for any monad is the application of `chain` to the composition of `of` and `fn`. Rambda did [it](https://github.com/ramda/ramda-fantasy/blob/master/src/Future.js#L33-L35).
To do it here one simply needs to move error checking from here to `run` method.

Reply:
Implementation of `map` through already existed `chain` and `of` is less efficient because of the creation of two Futures. That said such implementation reuses code and is the same for any monad. Though the duplicated code here is tiny meaning it may be allowed here for clarity.
So there is a tradeoff between performance and ideological purity here :-)